### PR TITLE
Folders: Able to fetch folders available for user as "shared" folder

### PR DESCRIPTION
**Pull request by alexanderzobnin**
**What is this feature?**

This PR is a base part extracted from #77345. We decided to go that way, but the actual representation is in discussion still. So we can merge the base functionality and figure out which UX solution might work better than other.

**Why do we need this feature?**

This is required for displaying all available folders in a tree when user does not have permissions to see all the root folders.

**Who is this feature for?**

[Add information on what kind of user the feature is for.]

**Which issue(s) does this PR fix?**:

Fixes #

Part of the #71425

**Special notes for your reviewer:**

This is the basic stuff, but we can add some performance improvements like cache in future PRs.

### CODE REVIEW

File: pkg/services/folder/folderimpl/folder.go [hunk 1]
```diff
@@ -246,6 +248,85 @@ func (s *Service) GetChildren(ctx context.Context, cmd *folder.GetChildrenQuery)
...
+		if f.ParentUID != "" {
+			nonRootFolders = append(nonRootFolders, f)
+		}
+	}
+
+	return nonRootFolders, nil
+}
+
+func (s *Service) deduplicateAvailableFolders(ctx context.Context, folders []*folder.Folder, rootFolders []*folder.Folder) []*folder.Folder {
+	allFolders := append(folders, rootFolders...)
```
Comment (alexanderzobnin) [inline pkg/services/folder/folderimpl/folder.go:298]: We need also to check if the folder with explicitly granted permission is a subfolder of one of the root folders that already available to user and remove that folders from view.

File: pkg/services/folder/folderimpl/folder.go [hunk 2]
```diff
@@ -246,6 +248,85 @@ func (s *Service) GetChildren(ctx context.Context, cmd *folder.GetChildrenQuery)
 	return filtered, nil
 }
 
+// GetSharedWithMe returns folders available to user, which cannot be accessed from the root folders
+func (s *Service) GetSharedWithMe(ctx context.Context, cmd *folder.GetChildrenQuery) ([]*folder.Folder, error) {
+	availableNonRootFolders, err := s.getAvailableNonRootFolders(ctx, cmd.OrgID, cmd.SignedInUser)
+	if err != nil {
+		return nil, folder.ErrInternal.Errorf("failed to fetch subfolders from dashboard store: %w", err)
```
Comment (papagian) [inline pkg/services/folder/folderimpl/folder.go:255]:
```suggestion
		return nil, folder.ErrInternal.Errorf("failed to fetch subfolders to which the user has explicit access: %w", err)
```

File: pkg/services/folder/folderimpl/folder.go [hunk 3]
```diff
@@ -246,6 +248,85 @@ func (s *Service) GetChildren(ctx context.Context, cmd *folder.GetChildrenQuery)
...
+	}
+	rootFolders, err := s.GetChildren(ctx, &folder.GetChildrenQuery{UID: "", OrgID: cmd.OrgID, SignedInUser: cmd.SignedInUser})
+	if err != nil {
+		return nil, folder.ErrInternal.Errorf("failed to fetch folders from dashboard store: %w", err)
```
Comment (papagian) [inline pkg/services/folder/folderimpl/folder.go:259]: nit; `s.GetChildren()` is a folder service function not a store one (also it fetches folders from the nested folders store not from dashboard store)
```suggestion
		return nil, folder.ErrInternal.Errorf("failed to fetch root folders to which the user has access: %w", err)
```

File: pkg/services/folder/folderimpl/folder.go (hunk 1 above)
Comment (papagian) [reply pkg/services/folder/folderimpl/folder.go:298]: do we need appending root folders given that any root folders to which the user has access are listed by the other endpoint?

Comment (alexanderzobnin) [reply pkg/services/folder/folderimpl/folder.go:298]: We only need to check against the root folder, so they are not added to the result. That's might be a strange behaviour when you have explicit access to some folder with access to it's parent folder on the root. In order to remove that folders we need to perform the check against root folders as well.

Comment (papagian) [reply pkg/services/folder/folderimpl/folder.go:298]: personally, I would prefer to list them twice instead of having to do that but I don't insist

**Review by papagian — APPROVED** [review]

**Pull request merged**